### PR TITLE
*: add TxnManager to manage txn in session

### LINKED ISSUE #30695: Introduce `TxnManager` and implement `GetTxnInfoSchema()` in it.

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #30695

### What is changed and how it works?

This PR belongs to refactoring txn in session: #30535 , including below works:

- Introduce a new package `sessiontxn` to manage transaction in session, we'll remove transaction related codes in session to this package later.
- Introduce a interface `TxnManager` to manage txn context in session. Currently in this PR, it only provides infoschema read method.
- Make `TxnManager` a interface to avoid dependency cycle, its real implement is another package `session` . We can use `sessiontxn.GetTxnManager(sctx)` to get it.
- `TxnManager.GetInfoShecma` is not used by other codes yet for make the pr as small as possible. But we added some failpoints to assert  `TxnManager.GetInfoShecma` equals to the values from the old implement.
- Add test cases to test the context provided are right for some scenes:
    - Some simple sql in implicit txn
    - Explicit txn
    - autocommit=0
    - Read-Committed isolation
    - @@tidb_snapshot
    - retry in optimistic transaction
    - lock conflict and retry in pessimistic transaction
    - stale read
    - prepare/execute
    - plan cache
    - some combined scenes, .i.e: stale read in prepare/execute

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by lcwangchao** [comment]
@cfzjywxk PTAL the transaction part
@Yisaer PTAL the stale read part

**Comment by lcwangchao** [comment]
@qw4990 PTAL plan cache part

File: sessiontxn/interface.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	"fmt"
+
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/stringutil"
+)
+
+// AssertRecordsKey is used to save failPoint invoke records
+// Only for test
```
Comment (tiancaiamao) [inline sessiontxn/interface.go:26]: If some variable is only used in the test code, you can write a xxx_test.go file, and then write

```
package sessiontxn

var AssertRecordsKey stringutil.StringerStr = "assertTxnManagerRecords"
```

in that file.

**Review by tiancaiamao — APPROVED** [review]
The actual changes is small, most of the code is testing related.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -474,6 +474,9 @@ type SessionVars struct {
 	//  TxnCtx Should be reset on transaction finished.
 	TxnCtx *TransactionContext
```
Comment (cfzjywxk) [inline sessionctx/variable/session.go:475]: Will this be removed in the future and all the information comes from the `txnManager` and related `provider`?

**Review by cfzjywxk — APPROVED** [review]

File: sessionctx/variable/session.go (hunk 2 above)
Comment (lcwangchao) [reply sessionctx/variable/session.go:475]: Yes, I think so. But It seems to be a big work and we need to be patient...

**Pull request merged**